src/main/java/net/neoforged/neoforge/network/registration/IPayloadRegistrar.java

```diff
@@ -0,0 +1,194 @@
+     */
+    <T extends CustomPacketPayload> IPayloadRegistrar configuration(ResourceLocation id, FriendlyByteBuf.Reader<T> reader, IConfigurationPayloadHandler<T> handler);
+
+    /**
+     * Registers a new payload type for the configuration phase.
+     * <p>
+     * This method allows different handlers to be registered for different packet-flows.
+     * <br>
+     * In practice this means that you can register a different handler for clientbound and serverbound packets,
+     * which allows you to handle them differently on the client and server side.
+     * </p>
+     *
+     * @param <T>     The type of the payload.
+     * @param id      The id of the payload.
+     * @param reader  The reader for the payload.
+     * @param handler The handler for the payload.
+     * @return The registrar.
+     * @implNote This method will capture all internal errors and wrap them in a {@link RegistrationFailedException}.
+     */
+    <T extends CustomPacketPayload> IPayloadRegistrar configuration(ResourceLocation id, FriendlyByteBuf.Reader<T> reader, Consumer<IDirectionAwarePayloadHandlerBuilder<T, IConfigurationPayloadHandler<T>>> handler);
+
+    /**
+     * Registers a new payload type for all supported phases.
+     *
+     * @param <T>     The type of the payload.
+     * @param id      The id of the payload.
+     * @param reader  The reader for the payload.
+     * @param handler The handler for the payload.
+     * @return The registrar.
+     */
+    default <T extends CustomPacketPayload> IPayloadRegistrar common(ResourceLocation id, FriendlyByteBuf.Reader<T> reader, IPayloadHandler<T> handler) {
+        return play(id, reader, handler::handle).configuration(id, reader, handler::handle);
+    }
+
+    /**
+     * Registers a new payload type for all supported phases.
+     * <p>
+     * This method allows different handlers to be registered for different packet-flows.
+     * <br>
+     * In practice this means that you can register a different handler for clientbound and serverbound packets,
+     * which allows you to handle them differently on the client and server side.
+     * </p>
+     *
+     * @param <T>     The type of the payload.
+     * @param id      The id of the payload.
+     * @param reader  The reader for the payload.
+     * @param handler The handler for the payload.
+     * @return The registrar.
+     */
+    <T extends CustomPacketPayload> IPayloadRegistrar common(ResourceLocation id, FriendlyByteBuf.Reader<T> reader, Consumer<IDirectionAwarePayloadHandlerBuilder<T, IPayloadHandler<T>>> handler);
+
+    /**
+     * Defines that the payloads registered by this registrar have a specific version associated with them.
+     * Clients connecting to a server with these payloads, will only be able to connect if they have the same version.
+     *
+     * @param version The version to use.
+     * @return The registrar, ready to configure payloads with that version.
```

Comment:
If the registration is immutable, it should be noted in this method that a new registrar is returned.